File: 103. Binary Tree Zigzag Level Order Traversal/103. Binary Tree Zigzag Level Order Traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function zigzagLevelOrder(root: TreeNode | null): number[][] {
    if (!root) return [];
    
    const result: number[][] = [];
    const queue: TreeNode[] = [root];
    let leftToRight: boolean = true;
    
    while (queue.length) {
        const levelSize: number = queue.length;
        const currentLevel: number[] = new Array(levelSize);
        
        for (let i = 0; i < levelSize; i++) {
            const node: TreeNode = queue.shift()!;
            const index: number = leftToRight ? i : levelSize - 1 - i;
            currentLevel[index] = node.val;
            
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
        
        leftToRight = !leftToRight;
        result.push(currentLevel);
    }
    
    return result;
}

File: 103. Binary Tree Zigzag Level Order Traversal/103. Binary Tree Zigzag Level Order Traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
def zigzag_level_order(root)
    return [] if root.nil?
    
    result = []
    queue = [root]
    left_to_right = true
    
    while !queue.empty?
        level_size = queue.size
        current_level = Array.new(level_size)
        
        level_size.times do |i|
            node = queue.shift
            index = left_to_right ? i : level_size - 1 - i
            current_level[index] = node.val
            
            queue << node.left if node.left
            queue << node.right if node.right
        end
        
        left_to_right = !left_to_right
        result << current_level
    end
    
    result
end